bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/FetchAndReserveCleaningStepService.kt

```diff
@@ -0,0 +1,198 @@
+    }
+
+    private fun taskResultsForErrors(errorTaskId: String, errors:  Collection<ErrorInfo<LegalEntityCreateError>>) : TaskStepResultEntryDto {
+
+        return TaskStepResultEntryDto( taskId = errorTaskId,errors = errors.map{TaskErrorDto(type= TaskErrorType.Unspecified,description = it.message)})
+    }
+
+    private fun legalEntityCreateTaskResult( task: TaskStepReservationEntryDto) : TaskStepResultEntryDto {
+
+        val legalEnity = task.businessPartner.legalEntity
+        val legalName = legalEnity?.legalName
+        val legalAddress = legalEnity?.legalAddress
+
+        return if ( legalEnity != null && legalName != null && legalAddress != null) {
+            val bpnLs = bpnIssuingService.issueLegalEntityBpns(1)
+            val legalEntityMetadataMap = metadataService.getMetadata(listOf(legalEnity)).toMapping()
+            val newLegalEntity= createLegalEntity(legalEnity, bpnLs[0], legalName, legalEntityMetadataMap)
+            // todo create address, changelogService and write to db
+            TaskStepResultEntryDto(
+                taskId = task.taskId,
+                businessPartner = BusinessPartnerFullDto(generic = task.businessPartner.generic)
+            )
+        } else{
+            TaskStepResultEntryDto( taskId = task.taskId,
+                    errors = listOf(TaskErrorDto(
+                            type= TaskErrorType.Unspecified,
+                            description = "Legal name or legal address is empty")
+                    )
+            )
+        }
+    }
+
+
+    private fun createLegalEntity(
+        legalEntityDto: LegalEntityDto,
+        bpnL: String,
+        legalNameValue: String,
+        metadataMap: BusinessPartnerBuildService.LegalEntityMetadataMapping
+    ): LegalEntity {
+
+        // it has to be validated that the legalForm exits
+        val legalForm = legalEntityDto.legalForm?.let { metadataMap.legalForms[it]!! }
+        val legalName = Name(
+            value = legalNameValue,
+            shortName = legalEntityDto.legalShortName
+        )
+        val newLegalEntity = LegalEntity(
+            bpn = bpnL,
+            legalName = legalName,
+            legalForm = legalForm,
+            currentness = Instant.now().truncatedTo(ChronoUnit.MICROS),
+        )
+        updateLegalEntity(newLegalEntity, legalEntityDto, legalName,legalEntityDto.identifiers.map { toEntity(it, metadataMap.idTypes, newLegalEntity) })
+
+        return newLegalEntity
+    }
+
+    private fun updateLegalEntity(
+        legalEntity: LegalEntity,
+        request: LegalEntityDto,
+        legalName: Name,
+        identifiers: List<LegalEntityIdentifier>
+    ) {
+
+        legalEntity.currentness = createCurrentnessTimestamp()
+
+        legalEntity.legalName = legalName
+
+        legalEntity.identifiers.clear()
+        legalEntity.states.clear()
+        legalEntity.classifications.clear()
+
+        legalEntity.states.addAll(request.states.map { toEntity(it, legalEntity) })
```

Comment:
There is a `.replace(..)` helper function that does `.clear()` and `.addAll()` in together.

Reply:
the whole method is refactored in the next step